File: static/js/prep/parse.js
// mjai event walking — twin of lib/parse.py for prep. Pure logic; no UI.
// Provides `flatten_mjai_log` and `walk_kyoku`, the canonical per-kyoku
// event walker that feeds defense / decision-state tracking on the JS side.
//
// See lib/parse.py for the documentation of the return shape — kept
// identical so the Python and JS prep layers can be diffed by a parity
// fixture verbatim.

(function (root, factory) {
  if (typeof module === "object" && module.exports) {
    module.exports = factory();
  } else {
    root.haipaiPrepParse = factory();
  }
}(typeof self !== "undefined" ? self : this, function () {

  function flatten_mjai_log(mjai_log) {
    const flat = [];
    for (const item of mjai_log || []) {
      if (Array.isArray(item)) {
        for (const sub of item) {
          if (sub && typeof sub === "object" && !Array.isArray(sub)) flat.push(sub);
        }
      } else if (item && typeof item === "object") {
        flat.push(item);
      }
    }
    return flat;
  }

  // True when any opponent's open-call count meets the shipped V2 open-threat
  // band at `turn` (1-based junme): 3 calls any turn, 2 from turn 7, 1 from
  // turn 11. Mirrors static/js/prep/defense.js::_is_open_threat (V2) — keep
  // the two in sync. A riichi opp can't have open melds, so it never trips
  // this; riichi precedence is handled in skill_area_for_entry regardless.
  function _open_threat_at(opponents, turn) {
    for (const seat in opponents) {
      const om = opponents[seat].open_melds || 0;
      if (om >= 3 || (turn >= 7 && om >= 2) || (turn >= 11 && om >= 1)) return true;
    }
    return false;
  }

  function walk_kyoku(events, start_pos, end_pos, player_id, target_tiles_left = 0) {
    const opponents = {};
    // First-seen seat order — Python's dict preserves insertion order; JS
    // objects with numeric-string keys do not, so we track it explicitly so
    // downstream defense consumers emit a stable seat ordering across runs.
    const opponent_order = [];
    const player_tsumo_riichi_state = [];
    // Parallel to player_tsumo_riichi_state, indexed by the player's junme:
    // true when a non-riichi opponent's open hand has tripped the open-threat
    // trigger at that turn. Feeds the Open Defense trends denominator the same
    // way player_tsumo_riichi_state feeds the riichi-Defense one.
    const open_threat_state = [];
    const reach_accepted_seats = new Set();
    const genbutsu_post_reach_by_seat = {};
    let tiles_left = 70;
    let first_dora_indicator = null;
    let bakaze = null;
    let oya = null;
    // Every tile that hit the table (dahai or kakan), in order. Together with
    // each opp's flow_pos_at_last_dahai this yields the open-threat genbutsu:
    // own discards ∪ flow since their last own dahai (temp-furiten window).
    const tile_flow = [];

    if (start_pos >= 0 && start_pos < events.length) {
      const sk = events[start_pos];
      if (sk && sk.type === "start_kyoku") {
        let dm = sk.dora_marker;
        if (Array.isArray(dm) && dm.length) dm = dm[0];
        if (typeof dm === "string") first_dora_indicator = dm;
        if (typeof sk.bakaze === "string") bakaze = sk.bakaze;
        if (typeof sk.oya === "number") oya = sk.oya;
      }
    }

    function ensureOpp(actor) {
      let o = opponents[actor];
      if (!o) {
        o = {
          discards: [],
          reach_event_idx: null,
          reach_accepted: false,
          open_melds: 0,
          melds: [],
          flow_pos_at_last_dahai: 0,
          ippatsu_alive: false,
        };
        opponents[actor] = o;
        opponent_order.push(actor);
      }
      return o;
    }

    for (let pos = start_pos + 1; pos < end_pos; pos++) {
      const e = events[pos];
      if (!e) continue;
      const etype = e.type;
      const actor = e.actor;

      if (etype === "tsumo") {
        tiles_left -= 1;
        if (actor === player_id) {
          player_tsumo_riichi_state.push(reach_accepted_seats.size > 0);
          // Same 1-based turn count defense.js uses for the open-threat band.
          const turn = player_tsumo_riichi_state.length;
          open_threat_state.push(_open_threat_at(opponents, turn));
        }
        // The riichi declarer's next draw ends their ippatsu window — one
        // full go-around has completed. The window survives the player's
        // and other non-riichi seats' draws.
        if (actor !== undefined && actor !== null && actor !== player_id) {
          const opp = opponents[actor];
          if (opp && opp.ippatsu_alive) opp.ippatsu_alive = false;
        }
      } else if (etype === "dahai" && actor !== undefined && actor !== null) {
        const pai = e.pai;
        if (pai !== undefined && pai !== null) tile_flow.push(pai);
        if (actor !== player_id) {
          const opp = ensureOpp(actor);
          opp.discards.push(pai);
          // Own dahai resets the temp-furiten window: only tiles after this
          // point count as "passed" for the open-threat genbutsu.
          opp.flow_pos_at_last_dahai = tile_flow.length;
        }
        if (pai !== undefined && pai !== null) {
          for (const seat of reach_accepted_seats) {
            if (!genbutsu_post_reach_by_seat[seat]) genbutsu_post_reach_by_seat[seat] = [];
            genbutsu_post_reach_by_seat[seat].push(pai);
          }
        }
      } else if (etype === "kakan" && actor !== undefined && actor !== null) {
        const pai = e.pai;
        if (pai !== undefined && pai !== null) tile_flow.push(pai);
        if (actor !== player_id) ensureOpp(actor);
        if (pai !== undefined && pai !== null) {
          for (const seat of reach_accepted_seats) {
            if (!genbutsu_post_reach_by_seat[seat]) genbutsu_post_reach_by_seat[seat] = [];
            genbutsu_post_reach_by_seat[seat].push(pai);
          }
        }
        // kakan interrupts an in-flight ippatsu window for every riichi opp.
        for (const seat of reach_accepted_seats) {
          const opp = opponents[seat];
          if (opp) opp.ippatsu_alive = false;
        }
      } else if (etype === "reach" && actor !== undefined && actor !== null && actor !== player_id) {
        const opp = ensureOpp(actor);
        opp.reach_event_idx = opp.discards.length;
      } else if (etype === "reach_accepted" && actor !== undefined && actor !== null && actor !== player_id) {
        const opp = ensureOpp(actor);
        opp.reach_accepted = true;
        opp.ippatsu_alive = true;
        reach_accepted_seats.add(actor);
      } else if ((etype === "pon" || etype === "chi" || etype === "daiminkan")
                 && actor !== undefined && actor !== null && actor !== player_id) {
        const opp = ensureOpp(actor);
        opp.open_melds += 1;
        const meld_tiles = (e.consumed || []).slice();
        if (e.pai !== undefined && e.pai !== null) meld_tiles.push(e.pai);
        opp.melds.push({ type: etype, tiles: meld_tiles });
        // Any open call by anyone breaks ippatsu for every riichi opponent.
        for (const seat of reach_accepted_seats) {
          const opp = opponents[seat];
          if (opp) opp.ippatsu_alive = false;
        }
      } else if ((etype === "pon" || etype === "chi" || etype === "daiminkan"
                  || etype === "ankan")
                 && actor !== undefined && actor !== null && actor === player_id) {
        // Player's own calls also break ippatsu for every riichi opp; ankan
        // by anyone other than the riichi declarer themselves is treated as
        // an interruption for the purposes of this signal.
        for (const seat of reach_accepted_seats) {
          const opp = opponents[seat];
          if (opp) opp.ippatsu_alive = false;
        }
      }

      if (tiles_left <= target_tiles_left) break;
    }

    return {
      opponents,
      opponent_order,
      player_tsumo_riichi_state,
      open_threat_state,
      genbutsu_post_reach_by_seat,
      first_dora_indicator,
      bakaze,
      oya,
      tile_flow,
      tiles_left_at_end: tiles_left,
    };
  }

  const _MELD_TYPES = new Set(["chi", "pon"]);
  const _KAN_TYPES = new Set(["ankan", "kakan", "daiminkan"]);

  // JS mirror of lib/parse.py::skill_area_for_entry — classifies one review
  // entry into the same denominator bucket the Python parser uses, so the
  // client-computed decision_counts agree with what the backfill would have
  // produced.
  // `open_threat` (optional) is true when a non-riichi opponent's open hand
  // tripped the open-threat trigger at this entry's turn. A plain dahai then
  // denominates against the Open Defense axis, mirroring how the categorizer
  // routes that scene to OD1/OD2/OD3. Riichi takes precedence over open, so
  // in_riichi is checked first. Python stays riichi-only (no open_defense);
  // the JS recompute path supplies the bucket for prepped games.
  function skill_area_for_entry(actual_type, expected_type, detail_types, in_riichi, open_threat) {
    const types = new Set();
    if (actual_type) types.add(actual_type);
    if (expected_type) types.add(expected_type);
    if (types.has("chi") || types.has("pon")) return "meld";
    if (types.has("reach")) return "riichi";
    if (types.has("ankan") || types.has("kakan") || types.has("daiminkan")) return "kan";
    if (types.has("dahai")) {
      if (in_riichi) return "defense";
      if (open_threat) return "open_defense";
      return "attack";
    }
    const d = new Set(detail_types || []);
    if (d.has("chi") || d.has("pon")) return "meld";
    if (d.has("reach")) return "riichi";
    if (d.has("ankan") || d.has("kakan") || d.has("daiminkan")) return "kan";
    return null;
  }

  // Per-skill-area decision counts for one kyoku — denominator for the
  // trends EV/D bars. Mirrors lib/parse.py::_decision_counts_for_kyoku;
  // entries hit attack/defense by player tsumo-state when the action is
  // plain dahai, otherwise by action-type priority.
  function decision_counts_for_kyoku(entries, start_pos, end_pos, events, player_id) {
    const counts = { attack: 0, defense: 0, open_defense: 0, riichi: 0, meld: 0, kan: 0 };
    const state = walk_kyoku(events, start_pos, end_pos, player_id);
    const junme_state = state.player_tsumo_riichi_state;
    const open_state = state.open_threat_state || [];
    for (const entry of (entries || [])) {
      const junme = entry.junme;
      const actual_type = (entry.actual || {}).type || null;
      const expected_type = (entry.expected || {}).type || null;
      const detail_types = (entry.details || []).map(d => (d.action || {}).type || null);
      const validJunme = typeof junme === "number" && junme >= 0;
      const in_riichi = validJunme && junme < junme_state.length && junme_state[junme];
      const open_threat = validJunme && junme < open_state.length && open_state[junme];
      const area = skill_area_for_entry(actual_type, expected_type, detail_types, in_riichi, open_threat);
      if (area) counts[area] += 1;
    }
    return counts;
  }

  // Roll per-kyoku denominators into a per-game total. Pass mortalData as
  // returned by /api/games/<id> (or a previously prepped game's
  // `mortal_data`). Returns null if the data isn't usable.
  function decision_counts_for_game(mortalData) {
    if (!mortalData) return null;
    const player_id = mortalData.player_id;
    if (player_id === undefined || player_id === null) return null;
    const kyokus = ((mortalData.review || {}).kyokus) || [];
    const events = flatten_mjai_log(mortalData.mjai_log || []);
    const start_positions = [];
    for (let i = 0; i < events.length; i++) {
      if (events[i] && events[i].type === "start_kyoku") start_positions.push(i);
    }
    const totals = { attack: 0, defense: 0, open_defense: 0, riichi: 0, meld: 0, kan: 0 };
    for (let ki = 0; ki < kyokus.length; ki++) {
      const start = start_positions[ki];
      if (start === undefined) continue;
      const end = (ki + 1 < start_positions.length) ? start_positions[ki + 1] : events.length;
      const c = decision_counts_for_kyoku(
        kyokus[ki].entries || [], start, end, events, player_id,
      );
      for (const k of Object.keys(totals)) totals[k] += c[k] || 0;
    }
    return totals;
  }

  return {
    flatten_mjai_log,
    walk_kyoku,
    skill_area_for_entry,
    decision_counts_for_kyoku,
    decision_counts_for_game,
  };
}));

File: lib/parse.py
#!/usr/bin/env python3
"""Parse Mortal AI JSON analysis into structured game data."""

from datetime import date
from typing import Any, NotRequired, TypedDict


class MistakeRecord(TypedDict):
    """A single mistake within a kyoku, as emitted by ``parse_game``.

    Categorize/DB layers later add fields (category, board_state, etc.) —
    those live in ``mistakes.data_json``.
    """
    turn: int
    ev_loss: float
    note: str | None
    hand: list[str]
    melds: list[Any]
    shanten: int | None
    draw: str | None
    actual: dict[str, Any]
    expected: dict[str, Any]
    top_actions: list[dict[str, Any]]


class RoundRecord(TypedDict):
    """One kyoku as emitted by ``parse_game``."""
    round: str              # e.g. "E1", "S3-1"
    honba: int
    turn_count: int
    decision_count: int
    decision_counts: dict[str, int] | None
    outcome: Any | None
    mistakes: list[MistakeRecord]


class GameRecord(TypedDict):
    """Return shape of ``parse_game``: structured game dict ready for
    ``db.add_game`` / ``compute_summary``.

    The DB layer later attaches ``mortal_file`` (set by the caller after
    saving the upload) and ``categorization_status``; the categorizer
    fills in per-mistake ``board_state`` / ``discard_stats`` / etc.
    """
    date: str
    log_url: str | None
    mortal_file: str | None
    rounds: list[RoundRecord]
    summary: dict[str, Any] | None
    # Optional fields layered on by callers / DB:
    game_id: NotRequired[int]
    categorization_status: NotRequired[str]


def severity(ev_loss):
    if ev_loss > 1.00:
        return "???"
    elif ev_loss >= 0.50:
        return "??"
    else:
        return "?"


def round_header(start):
    """Build round header string from a start_kyoku event."""
    header = f"{start['bakaze']}{start['kyoku']}"
    if start["honba"] > 0:
        header += f"-{start['honba']}"
    return header


def format_action(action):
    """Format an action dict for text display."""
    t = action.get("type", "?")
    if t == "dahai":
        return action["pai"]
    elif t in ("chi", "pon"):
        consumed = "".join(action.get("consumed", []))
        return f"{t} {consumed}+{action.get('pai', '?')}"
    elif t == "reach":
        return "riichi"
    elif t == "hora":
        return "win"
    elif t == "none":
        return "pass"
    elif t == "ankan":
        return f"ankan {action.get('consumed', ['?'])[0]}"
    return t


def flatten_mjai_log(mjai_log):
    """mjai_log entries can be lists of simultaneous events; flatten to a
    single sequential list of event dicts."""
    flat = []
    for item in mjai_log:
        if isinstance(item, dict):
            flat.append(item)
        elif isinstance(item, list):
            for sub in item:
                if isinstance(sub, dict):
                    flat.append(sub)
    return flat


def walk_kyoku(events, start_pos, end_pos, player_id, target_tiles_left=0):
    """Single canonical pass over a kyoku's mjai events.

    Feeds every defense / attack-side extractor: classic-suji walker
    (``lib/defense.py``), KillerDucky walker (``lib/defense_kd.py``), and
    parse's per-junme decision-state tracker (CS-02). Walking once per
    consumer call is fine — the win is *one implementation* of "what did
    each opponent do, and was anyone in riichi at each player tsumo".

    Walks ``events[start_pos+1 : end_pos]`` and stops as soon as
    ``tiles_left <= target_tiles_left`` (defense walkers cut at the
    mistake's own tsumo; ``parse_game`` walks the full kyoku with the
    default ``0``).

    Returns a dict::

        {
          "opponents": {seat: {
              "discards": [mjai, ...],   # raw, includes red-five marks
              "reach_event_idx": int|None,
                  # len(discards) AT the moment the `reach` event fired —
                  # i.e. the slot the riichi tile WILL land on at the next
                  # dahai. Adapters that want the historic off-by-one
                  # semantic subtract 1; KD-style adapters use it as is
                  # (and check `< len(discards)` to confirm the riichi
                  # tile actually got discarded).
              "reach_accepted": bool,    # `reach_accepted` event observed
              "open_melds": int,         # pon/chi/daiminkan count (kakan
                                         # doesn't bump — already counted
                                         # by the underlying pon)
          }},
          "player_tsumo_riichi_state": [bool, ...],
              # one entry per player tsumo, in order. True iff at least
              # one opponent had already been reach_accepted. Indexable
              # by junme.
          "genbutsu_post_reach_by_seat": {seat: [mjai, ...]},
              # for each opp seat that ever hit reach_accepted: every
              # dahai/kakan tile (any actor) seen after that moment.
              # Raw mjai (red-five marks preserved). KD adapter
              # normalises and unions with the seat's own discards.
          "first_dora_indicator": mjai|None,
          "tiles_left_at_end": int,
        }
    """
    opponents = {}
    player_tsumo_riichi_state = []
    reach_accepted_seats = set()
    genbutsu_post_reach_by_seat = {}
    tiles_left = 70
    first_dora_indicator = None

    if 0 <= start_pos < len(events):
        sk = events[start_pos]
        if sk.get("type") == "start_kyoku":
            dm = sk.get("dora_marker")
            if isinstance(dm, list) and dm:
                dm = dm[0]
            if isinstance(dm, str):
                first_dora_indicator = dm

    def _ensure(actor):
        return opponents.setdefault(actor, {
            "discards": [],
            "reach_event_idx": None,
            "reach_accepted": False,
            "open_melds": 0,
        })

    for pos in range(start_pos + 1, end_pos):
        e = events[pos]
        etype = e.get("type")
        actor = e.get("actor")

        if etype == "tsumo":
            tiles_left -= 1
            if actor == player_id:
                player_tsumo_riichi_state.append(bool(reach_accepted_seats))

        elif etype == "dahai" and actor is not None:
            pai = e.get("pai")
            if actor != player_id:
                _ensure(actor)["discards"].append(pai)
            if pai is not None:
                for seat in reach_accepted_seats:
                    genbutsu_post_reach_by_seat.setdefault(seat, []).append(pai)

        elif etype == "kakan" and actor is not None:
            pai = e.get("pai")
            # kakan upgrades an existing pon — register the seat (so the
            # opp dict isn't missing) but don't bump open_melds.
            if actor != player_id:
                _ensure(actor)
            if pai is not None:
                for seat in reach_accepted_seats:
                    genbutsu_post_reach_by_seat.setdefault(seat, []).append(pai)

        elif etype == "reach" and actor is not None and actor != player_id:
            opp = _ensure(actor)
            opp["reach_event_idx"] = len(opp["discards"])

        elif etype == "reach_accepted" and actor is not None and actor != player_id:
            opp = _ensure(actor)
            opp["reach_accepted"] = True
            reach_accepted_seats.add(actor)

        elif etype in ("pon", "chi", "daiminkan") and actor is not None and actor != player_id:
            _ensure(actor)["open_melds"] += 1

        if tiles_left <= target_tiles_left:
            break

    return {
        "opponents": opponents,
        "player_tsumo_riichi_state": player_tsumo_riichi_state,
        "genbutsu_post_reach_by_seat": genbutsu_post_reach_by_seat,
        "first_dora_indicator": first_dora_indicator,
        "tiles_left_at_end": tiles_left,
    }


_MELD_TYPES = frozenset({"chi", "pon"})
_KAN_TYPES = frozenset({"ankan", "kakan", "daiminkan"})


def skill_area_for_entry(actual_type, expected_type, detail_types=(),
                         in_riichi=False):
    """Classify a Mortal review entry into exactly one skill area.

    Returns ``"attack"``, ``"defense"``, ``"meld"``, ``"riichi"``, ``"kan"``,
    or ``None`` (entry isn't a trackable player decision).

    Used as the denominator bucket for the trends EV/D bars: each entry
    lands in exactly one area, the highest-priority non-dahai action type
    in either ``actual.type`` or ``expected.type``. Plain dahai falls
    through to attack/defense based on whether the player is in riichi.
    """
    types = {actual_type, expected_type}
    if types & _MELD_TYPES:
        return "meld"
    if "reach" in types:
        return "riichi"
    if types & _KAN_TYPES:
        return "kan"
    if "dahai" in types:
        return "defense" if in_riichi else "attack"
    d = set(detail_types)
    if d & _MELD_TYPES:
        return "meld"
    if "reach" in d:
        return "riichi"
    if d & _KAN_TYPES:
        return "kan"
    return None


def _decision_counts_for_kyoku(entries, start_pos, end_pos, events, player_id):
    """Per-skill-area decision counts for one kyoku — the denominators used
    by the trends EV/D bars.

    Each entry lands in exactly ONE bucket (sum of buckets = number of
    counted entries), via ``skill_area_for_entry``. That shares its
    priority rule with ``categorize_by_action_type`` / CATEGORY_INFO
    groups, so a mistake's skill area (via its code's group) and its
    denominator bucket always agree: a 5B (missed riichi) goes to the
    ``riichi`` denominator, a D2 push goes to ``defense``, etc.
    """
    counts = {"attack": 0, "defense": 0, "riichi": 0, "meld": 0, "kan": 0}
    state = walk_kyoku(events, start_pos, end_pos, player_id)
    junme_riichi_state = state["player_tsumo_riichi_state"]

    for entry in entries:
        junme = entry.get("junme")
        actual_type = (entry.get("actual") or {}).get("type")
        expected_type = (entry.get("expected") or {}).get("type")
        detail_types = [
            (d.get("action") or {}).get("type")
            for d in entry.get("details", [])
        ]
        in_riichi = (
            junme is not None
            and 0 <= junme < len(junme_riichi_state)
            and junme_riichi_state[junme]
        )
        area = skill_area_for_entry(
            actual_type, expected_type, detail_types, in_riichi=in_riichi
        )
        if area:
            counts[area] += 1
    return counts


def parse_game(data, game_date=None) -> GameRecord:
    """Parse Mortal JSON into a structured ``GameRecord``.

    Returns a ``GameRecord`` (TypedDict, runtime ``dict``) with keys
    ``date``, ``log_url``, ``mortal_file``, ``rounds``, ``summary``.
    The summary and annotations are layered on by the DB / categorizer
    downstream. See ``GameRecord`` / ``RoundRecord`` / ``MistakeRecord``
    declarations at the top of this module for the full shape.
    """
    game_date = game_date or date.today().isoformat()

    if not isinstance(data, dict):
        raise ValueError("Expected a JSON object, got " + type(data).__name__)
    if "review" not in data or not isinstance(data.get("review"), dict):
        raise ValueError("Missing or invalid 'review' field")
    if "kyokus" not in data["review"] or not isinstance(data["review"]["kyokus"], list):
        raise ValueError("Missing or invalid 'review.kyokus' field")
    if "mjai_log" not in data or not isinstance(data.get("mjai_log"), list):
        raise ValueError("Missing or invalid 'mjai_log' field")

    kyokus = data["review"]["kyokus"]
    start_events = [
        e for e in data["mjai_log"]
        if isinstance(e, dict) and e.get("type") == "start_kyoku"
    ]

    if len(kyokus) != len(start_events):
        raise ValueError(
            f"{len(kyokus)} review kyokus but {len(start_events)} start_kyoku events"
        )

    flat_events = flatten_mjai_log(data["mjai_log"])
    start_positions = [
        i for i, e in enumerate(flat_events) if e.get("type") == "start_kyoku"
    ]
    player_id = data.get("player_id")

    rounds = []
    for kyoku_idx, (kyoku, start) in enumerate(zip(kyokus, start_events)):
        entries = kyoku["entries"]
        turn_count = (max(e["junme"] for e in entries) + 1) if entries else 0
        decision_count = len(entries)
        start_pos = start_positions[kyoku_idx] if kyoku_idx < len(start_positions) else None
        end_pos = (start_positions[kyoku_idx + 1]
                   if start_pos is not None and kyoku_idx + 1 < len(start_positions)
                   else len(flat_events))
        decision_counts = (
            _decision_counts_for_kyoku(entries, start_pos, end_pos, flat_events, player_id)
            if start_pos is not None and player_id is not None
            else None
        )

        mistakes = []
        for entry in entries:
            if not entry["is_equal"]:
                try:
                    expected_q = entry["details"][0]["q_value"]
                    actual_q = entry["details"][entry["actual_index"]]["q_value"]
                except (KeyError, IndexError, TypeError) as e:
                    raise ValueError(f"Malformed entry in kyoku: {e}") from e
                ev_loss = round(expected_q - actual_q, 2)

                # Store all Mortal-evaluated actions so the EV table can show
                # a Q-value for every highlighted tile (actual/expected/best_discard)
                # without the "-" dash that appeared whenever the third
                # highlight row wasn't in Mortal's top 3.
                top_actions = [
                    {
                        "action": d["action"],
                        "q_value": round(d["q_value"], 4),
                        "prob": round(d["prob"], 4),
                    }
                    for d in entry["details"]
                ]

                mistakes.append({
                    "turn": entry["junme"],
                    "ev_loss": ev_loss,
                    "note": None,
                    "hand": entry["state"]["tehai"],
                    "melds": entry["state"]["fuuros"],
                    "shanten": entry.get("shanten"),
                    "draw": entry.get("tile"),
                    "actual": entry["actual"],
                    "expected": entry["expected"],
                    "top_actions": top_actions,
                })

        rounds.append({
            "round": round_header(start),
            "honba": start["honba"],
            "turn_count": turn_count,
            "decision_count": decision_count,
            "decision_counts": decision_counts,
            "outcome": None,
            "mistakes": mistakes,
        })

    return {
        "date": game_date,
        "log_url": None,
        "mortal_file": None,
        "rounds": rounds,
        "summary": None,
    }


